pkg/common/cns-lib/volume/manager.go

```diff
@@ -184,161 +192,235 @@ func (m *defaultManager) ResetManager(ctx context.Context, vcenter *cnsvsphere.V
-		var task *object.Task
-		var taskInfo *vim25types.TaskInfo
-		// Store the volume name passed in by input spec, this name may exceed 80 characters.
-		volNameFromInputSpec := spec.Name
-		// Call the CNS CreateVolume.
-		taskDetailsInMap, ok := volumeTaskMap[volNameFromInputSpec]
-		if ok {
-			task = taskDetailsInMap.task
-			log.Infof("CreateVolume task still pending for Volume: %q, with taskInfo: %+v",
-				volNameFromInputSpec, task)
-		} else {
-			// Truncate the volume name to make sure the name is within 80 characters before calling CNS.
-			if len(spec.Name) > maxLengthOfVolumeNameInCNS {
-				volNameAfterTruncate := spec.Name[0 : maxLengthOfVolumeNameInCNS-1]
-				spec.Name = volNameAfterTruncate
-				log.Infof("Create Volume with name %s is too long, truncate it to %s", volNameFromInputSpec, spec.Name)
-				log.Debugf("CNS Create Volume is called with %v", spew.Sdump(*spec))
-			}
-			cnsCreateSpecList = append(cnsCreateSpecList, *spec)
-			task, err = m.virtualCenter.CnsClient.CreateVolume(ctx, cnsCreateSpecList)
+		if !isStaticallyProvisioned(spec) {
+			// Persist task details only for dynamically provisioned volumes.
+			volumeOperationDetails = createRequestDetails(volNameFromInputSpec, "", "", 0,
+				volumeOperationDetails.OperationDetails.TaskInvocationTimestamp,
+				task.Reference().Value, "", taskInvocationStatusInProgress, "")
+			err = m.operationStore.StoreRequestDetails(ctx, volumeOperationDetails)
 			if err != nil {
-				log.Errorf("CNS CreateVolume failed from vCenter %q with err: %v", m.virtualCenter.Config.Host, err)
-				return nil, err
-			}
-			var isStaticallyProvisionedBlockVolume bool
-			var isStaticallyProvisionedFileVolume bool
-			if spec.VolumeType == string(cnstypes.CnsVolumeTypeBlock) {
-				blockBackingDetails, ok := spec.BackingObjectDetails.(*cnstypes.CnsBlockBackingDetails)
-				if ok && (blockBackingDetails.BackingDiskId != "" || blockBackingDetails.BackingDiskUrlPath != "") {
-					isStaticallyProvisionedBlockVolume = true
-				}
-			}
-			if spec.VolumeType == string(cnstypes.CnsVolumeTypeFile) {
-				fileBackingDetails, ok := spec.BackingObjectDetails.(*cnstypes.CnsVsanFileShareBackingDetails)
-				if ok && fileBackingDetails.BackingFileId != "" {
-					isStaticallyProvisionedFileVolume = true
-				}
-			}
-			// Add the task details to volumeTaskMap only for dynamically
-			// provisioned volumes. For static volume provisioning we need not
-			// store the taskDetails as it doesn't result in orphaned volumes.
-			if !isStaticallyProvisionedBlockVolume && !isStaticallyProvisionedFileVolume {
-				var taskDetails createVolumeTaskDetails
-				// Store task details and expiration time in volumeTaskMap.
-				taskDetails.task = task
-				taskDetails.expirationTime = time.Now().Add(time.Hour * time.Duration(defaultOpsExpirationTimeInHours))
-				volumeTaskMap[volNameFromInputSpec] = &taskDetails
+				// Don't return if CreateVolume details can't be stored.
```

Comment:
What is the effect of not storing the volumeOperationDetails?

Reply:
There are two cases:
1. We return here with an error. Since the details aren't stored, the subsequent CSI CreateVolume request is going to invoke another CNS CreateVolume request and we'd have an orphan volume.
2. We continue with the execution. Its possible that we dont store the details on the API server, but CNS succeeds and we return success. K8s will not retry the CreateVolume operation so we wont have an orphan volume.

Reply:
On further thought, on line 256 in this PR, the `defer` function only stores details if the status is `Success` or `Error`. Do you think its worth modifying that logic to the following:
```
Store the details on the API server if:
1. Status is Success or Error OR
2. Status is in progress but the previous attempt to store the details (over here - line 285) failed
```

Reply:
I guess my concern was whether we will have a case where we do not have the CR but the CSI operation is successful. Its easy to assume that we always have a CR to know what happened to the CSI calls during debugging, but we have this case here where we just ignore the CR add/update.

Reply:
After discussion offline, we will keep the behaviour as is. FVT and ST teams will get back to us, after extensive testing, about the frequency of orphan volumes being created and API server updates failing. This was called out in the functional spec review with these teams.